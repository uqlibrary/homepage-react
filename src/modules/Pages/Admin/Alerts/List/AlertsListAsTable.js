import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useCookies } from 'react-cookie';
import Chip from '@mui/material/Chip';
import Checkbox from '@mui/material/Checkbox';
import Grid from '@mui/material/Grid';
import { InlineLoader } from 'modules/SharedComponents/Toolbox/Loaders';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableFooter from '@mui/material/TableFooter';
import TableHead from '@mui/material/TableHead';
import TablePagination from '@mui/material/TablePagination';
import TableRow from '@mui/material/TableRow';
import Paper from '@mui/material/Paper';
import IconButton from '@mui/material/IconButton';

import CloseIcon from '@mui/icons-material/Close';
import DeleteIcon from '@mui/icons-material/Delete';

import { TablePaginationActions } from './TablePaginationActions';
import { ConfirmationBox } from 'modules/SharedComponents/Toolbox/ConfirmDialogBox';
import { useConfirmationState } from 'hooks';
import { default as locale } from '../alertsadmin.locale';
import AlertSplitButton from './AlertSplitButton';
import { systemList } from '../alerthelpers';
import { scrollToTopOfPage } from 'helpers/general';
import { styled } from '@mui/material/styles';
import { useNavigate } from 'react-router-dom';

const moment = require('moment');

// original based on https://codesandbox.io/s/hier2
// per https://material-ui.com/components/tables/#custom-pagination-actions

const StyledHeader = styled('div')(({ theme }) => ({
    display: 'flex',
    padding: '0 0.5rem',
    '&.headerRowHighlighted': {
        backgroundColor: theme.palette.primary.main,
        color: '#fff',
    },
    '& .iconHighlighted': {
        color: '#fff',
    },
}));

const StyledTable = styled(Table)(({ theme }) => ({
    '&.table': {
        minWidth: 500,
    },
    '& .startDate': {
        whiteSpace: 'pre', // makes moment format able to take a carriage return
    },
    '& .endDate': {
        whiteSpace: 'pre',
    },
    '& .chipblock': {
        '&>div': {
            marginBottom: 4,
        },
        '&>div>div': {
            marginBottom: 4,
        },
    },
    '& .urgent': {
        backgroundColor: theme.palette.warning.light,
        color: '#000',
    },
    '& .extreme': {
        backgroundColor: theme.palette.error.main,
        color: '#fff',
    },
    '& .system': {
        backgroundColor: '#666666',
        color: '#fff',
    },
    '& .checkboxCell': {
        '& input[type="checkbox"]:checked + svg': {
            fill: '#595959',
        },
    },
    '& .removedChip': {
        textDecoration: 'line-through',
    },
    '& .screenreader': {
        position: 'absolute',
        left: -9999,
    },
}));

export const AlertsListAsTable = ({
    rows,
    headertag,
    alertsLoading,
    actions,
    deleteAlert,
    footerDisplayMinLength,
    alertOrder,
}) => {
    const navigate = useNavigate();
    const [page, setPage] = useState(0);
    const [deleteActive, setDeleteActive] = useState(false);
    const [alertNotice, setAlertNotice] = useState('');
    const [cookies, setCookie] = useCookies();

    const [rowsPerPage, setRowsPerPage] = useState(
        (!cookies.alertAdminPaginatorSize && footerDisplayMinLength) || parseInt(cookies.alertAdminPaginatorSize, 10),
    );

    const [isDeleteConfirmOpen, showDeleteConfirmation, hideDeleteConfirmation] = useConfirmationState();
    const [
        isDeleteFailureConfirmationOpen,
        showDeleteFailureConfirmation,
        hideDeleteFailureConfirmation,
    ] = useConfirmationState();

    React.useEffect(() => {
        // make it redraw when all displayed rows in a table are deleted
        setPage(0);
    }, [rows]);

    const handleChangePage = (event, newPage) => {
        setPage(newPage);
    };

    const checkBoxIdPrefix = 'checkbox-';

    const handleChangeRowsPerPage = event => {
        const numberOfRows = parseInt(event.target.value, 10);

        const current = new Date();
        const nextYear = new Date();
        nextYear.setFullYear(current.getFullYear() + 1);
        setCookie('alertAdminPaginatorSize', numberOfRows, { expires: nextYear });

        setRowsPerPage(numberOfRows);
        setPage(0);
    };

    const tableType = headertag.replace(' alerts', '').toLowerCase();

    const headerCountIndicator = '[N] alert[s]'.replace('[N]', rows.length).replace('[s]', rows.length > 1 ? 's' : '');

    let userows = rows;
    if (!!alertOrder && !!rows && rows.length > 0) {
        // (fake switch to allow istanbul on else if :( )
        switch (alertOrder) {
            case 'reverseEnd':
                userows = rows.sort(
                    (a, b) => moment(b.end, 'YYYY-MM-DD hh:mm:ss') - moment(a.end, 'YYYY-MM-DD hh:mm:ss'),
                );
                break;
            /* istanbul ignore next */
            case 'forwardEnd':
                userows = rows.sort(
                    (a, b) => moment(a.end, 'YYYY-MM-DD hh:mm:ss') - moment(b.end, 'YYYY-MM-DD hh:mm:ss'),
                );
                break;
            /* istanbul ignore next */
            case 'forwardStart':
                userows = rows.sort(
                    (a, b) => moment(a.start, 'YYYY-MM-DD hh:mm:ss') - moment(b.start, 'YYYY-MM-DD hh:mm:ss'),
                );
                break;
            /* istanbul ignore next */
            default:
        }
    }

    if (!!alertsLoading) {
        return (
            <Grid
                item
                xs={'auto'}
                style={{
                    width: 80,
                    marginRight: 20,
                    marginBottom: 6,
                    opacity: 0.3,
                    height: 200, // default to some space for the blocks
                }}
            >
                <InlineLoader message="Loading" />
            </Grid>
        );
    }

    const navigateToEditForm = alertid => {
        navigate(`/admin/alerts/edit/${alertid}`);
        scrollToTopOfPage();
    };

    const navigateToCloneForm = alertid => {
        navigate(`/admin/alerts/clone/${alertid}`);
        scrollToTopOfPage();
    };

    const navigateToView = alertid => {
        navigate(`/admin/alerts/view/${alertid}`);
        scrollToTopOfPage();
    };

    /* istanbul ignore next */
    const reEnableAllCheckboxes = () => {
        // is tested but coverage failing
        const checkBoxList = document.querySelectorAll('#admin-alerts-list input[type="checkbox"]');
        checkBoxList.forEach(ii => {
            ii.disabled = false;
            ii.parentElement.parentElement.classList.remove('Mui-disabled');
        });
    };

    /* istanbul ignore next */
    const clearAllCheckboxes = () => {
        // is tested but coverage failing
        const checkBoxList = document.querySelectorAll('#admin-alerts-list input[type="checkbox"]');
        checkBoxList.forEach(ii => {
            if (ii.checked) {
                ii.click();
            }
        });
    };

    function getNumberCheckboxesSelected() {
        return document.querySelectorAll('#admin-alerts-list tr.alert-data-row :checked').length;
    }

    /* istanbul ignore next */
    const handleCheckboxChange = e => {
        // is tested but coverage failing
        const numberCheckboxesSelected = getNumberCheckboxesSelected();

        const thisType = e.target.closest('table').parentElement.id;
        /* istanbul ignore else */
        if (!!e.target && !!e.target.checked) {
            // handle a checkbox being turned on
            if (numberCheckboxesSelected === 1) {
                setDeleteActive(true);
            }
            // disable any checkboxes in a different alert list
            const checkBoxList = document.querySelectorAll('#admin-alerts-list input[type="checkbox"]');
            checkBoxList.forEach(ii => {
                const thetype = ii.closest('table').parentElement.id;
                if (thetype !== thisType) {
                    ii.disabled = true;
                    ii.parentElement.parentElement.classList.add('Mui-disabled');
                }
            });
        } /* istanbul ignore else */ else if (!!e.target && !e.target.checked) {
            // handle a checkbox being turned off
            if (numberCheckboxesSelected === 0) {
                setDeleteActive(false);
                reEnableAllCheckboxes();
            }
        }
        setAlertNotice(
            '[n] alert[s] selected'
                .replace('[n]', numberCheckboxesSelected)
                .replace('[s]', numberCheckboxesSelected === 1 ? '' : 's'),
        );
    };

    /* istanbul ignore next */
    function deleteAlertById(alertID) {
        // this is called, but the coverage isn't happening
        deleteAlert(alertID)
            .then(() => {
                setAlertNotice('');
                setDeleteActive(false);
                actions.loadAllAlerts();
            })
            .catch(() => {
                showDeleteFailureConfirmation();
            });
    }

    /* istanbul ignore next */
    const deleteSelectedAlerts = () => {
        // is tested but coverage failing
        const checkboxes = document.querySelectorAll('#admin-alerts-list input[type="checkbox"]:checked');
        /* istanbul ignore else */
        if (!!checkboxes && checkboxes.length > 0) {
            checkboxes.forEach(c => {
                const alertID = c.value.replace(checkBoxIdPrefix, '');
                deleteAlertById(alertID);
            });
        }
    };

    const confirmDeleteLocale = numberOfCheckedBoxes => {
        return {
            ...locale.listPage.confirmDelete,
            confirmationTitle: locale.listPage.confirmDelete.confirmationTitle
                .replace('[N]', numberOfCheckedBoxes)
                .replace('alerts', numberOfCheckedBoxes === 1 ? 'alert' : 'alerts'),
        };
    };

    const needsPaginator = userows.length > footerDisplayMinLength;
    return (
        <React.Fragment>
            <ConfirmationBox
                actionButtonColor="secondary"
                actionButtonVariant="contained"
                confirmationBoxId="alert-delete-confirm"
                onAction={deleteSelectedAlerts}
                onClose={hideDeleteConfirmation}
                onCancelAction={hideDeleteConfirmation}
                isOpen={isDeleteConfirmOpen}
                locale={confirmDeleteLocale(getNumberCheckboxesSelected())}
            />
            <ConfirmationBox
                actionButtonColor="primary"
                actionButtonVariant="contained"
                confirmationBoxId="alert-delete-error-dialog"
                onAction={hideDeleteFailureConfirmation}
                onClose={hideDeleteFailureConfirmation}
                hideCancelButton
                isOpen={isDeleteFailureConfirmationOpen}
                locale={locale.listPage.deleteError}
            />
            <StyledHeader
                data-testid={`headerRow-${tableType}`}
                className={`${!!deleteActive ? /* istanbul ignore next */ 'headerRowHighlighted' : ''}`}
            >
                <div>
                    <h3 style={{ marginBottom: 6 }}>
                        {headertag}
                        <span
                            style={{ fontSize: '0.9em', fontWeight: 300 }}
                            data-testid={`headerRow-count-${tableType}`}
                        >
                            {' '}
                            - {headerCountIndicator}
                        </span>
                    </h3>
                </div>
                {!!deleteActive && (
                    /* istanbul ignore next */ <span
                        className="deleteManager"
                        style={{ marginLeft: 'auto', paddingTop: 8 }}
                    >
                        <span>{alertNotice}</span>
                        <IconButton
                            onClick={showDeleteConfirmation}
                            aria-label="Delete alert(s)"
                            data-testid={`alert-list-${tableType}-delete-button`}
                            title="Delete alert(s)"
                            size="large"
                        >
                            <DeleteIcon
                                className={`${
                                    !!deleteActive
                                        ? /* istanbul ignore next */ 'iconHighlighted'
                                        : /* istanbul ignore next */ ''
                                }`}
                            />
                        </IconButton>
                        <IconButton
                            onClick={clearAllCheckboxes}
                            aria-label="Deselect all"
                            data-testid={`alert-list-${tableType}-deselect-button`}
                            className={'iconHighlighted'}
                            title="Deselect all"
                            size="large"
                        >
                            <CloseIcon />
                        </IconButton>
                    </span>
                )}
            </StyledHeader>
            <TableContainer id={`alert-list-${tableType}`} data-testid={`alert-list-${tableType}`} component={Paper}>
                <StyledTable className={'table'} aria-label="custom pagination table" style={{ minHeight: 200 }}>
                    <TableHead>
                        <TableRow md-row="" className="md-row">
                            <TableCell component="th" scope="row" align="center">
                                <span className={'screenreader'}>Select</span>
                            </TableCell>
                            <TableCell component="th" scope="row">
                                Alert
                            </TableCell>
                            <TableCell component="th" scope="row" align="center">
                                Publish date
                            </TableCell>
                            <TableCell component="th" scope="row" align="center">
                                Unpublish date
                            </TableCell>
                            <TableCell component="th" scope="row" align="center">
                                <span className={'screenreader'}>Actions</span>
                            </TableCell>
                        </TableRow>
                    </TableHead>
                    <TableBody>
                        {rowsPerPage > 0 &&
                            userows.length > 0 &&
                            userows.slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage).map(alert => {
                                return (
                                    <TableRow
                                        key={alert.id}
                                        id={`alert-list-row-${alert.id}`}
                                        data-testid={`alert-list-row-${alert.id}`}
                                        className="alert-data-row"
                                    >
                                        <TableCell component="td" className={'checkboxCell'}>
                                            <Checkbox
                                                id={`alert-list-item-checkbox-${alert.id}`}
                                                inputProps={{
                                                    'aria-labelledby': `alert-list-item-title-${alert.id}`,
                                                    'data-testid': `alert-list-item-checkbox-${alert.id}`,
                                                }}
                                                onChange={handleCheckboxChange}
                                                value={`${checkBoxIdPrefix}${alert.id}`}
                                            />
                                        </TableCell>
                                        <TableCell component="td" className="alertText">
                                            <h4
                                                style={{ display: 'inline' }}
                                                id={`alert-list-item-title-${alert.id}`}
                                            >{`${alert.title}`}</h4>{' '}
                                            {`${alert.message.replace('[permanent]', '')}`}
                                            <div className={'chipblock'}>
                                                {!!alert && alert.priority_type === 'urgent' && (
                                                    <Chip
                                                        data-testid={`alert-list-urgent-chip-${alert.id}`}
                                                        label="Urgent"
                                                        title={locale.form.tooltips.priority.level.urgent}
                                                        className={'urgent'}
                                                    />
                                                )}{' '}
                                                {!!alert && alert.priority_type === 'extreme' && (
                                                    <Chip
                                                        data-testid={`alert-list-urgent-chip-${alert.id}`}
                                                        label="Extreme"
                                                        title={locale.form.tooltips.priority.level.extreme}
                                                        className={'extreme'}
                                                    />
                                                )}{' '}
                                                {alert.body.includes('](') && (
                                                    <Chip
                                                        data-testid={`alert-list-link-chip-${alert.id}`}
                                                        label="Link"
                                                        title="This alert has a link out"
                                                    />
                                                )}{' '}
                                                {alert.body.includes('[permanent]') && (
                                                    <Chip
                                                        data-testid={`alert-list-permanent-chip-${alert.id}`}
                                                        label="Permanent"
                                                        title="This alert cannot be dismissed"
                                                    />
                                                )}
                                                {!!alert.systems &&
                                                    alert.systems.length > 0 &&
                                                    alert.systems.map((systemSlug, index) => {
                                                        const ss = systemList.find(s => s.slug === systemSlug);
                                                        return (
                                                            <div
                                                                key={`alert-list-system-chip-${ss?.slug ||
                                                                    /* istanbul ignore next */ index}`}
                                                                style={{ marginLeft: 3, display: 'inline' }}
                                                            >
                                                                <Chip
                                                                    data-testid={`alert-list-system-chip-${
                                                                        alert.id
                                                                    }-${ss?.slug || /* istanbul ignore next */ index}`}
                                                                    label={`System: ${ss?.title ||
                                                                        /* istanbul ignore next */ ss?.slug ||
                                                                        /* istanbul ignore next */ 'Unrecognised'}`}
                                                                    title={`This alert is restricted to the ${ss?.title ||
                                                                        /* istanbul ignore next */ ss?.slug ||
                                                                        /* istanbul ignore next */ 'Unrecognised'} system`}
                                                                    className={`system ${!!ss?.removed &&
                                                                        /* istanbul ignore next */ 'removedChip'}`}
                                                                />
                                                            </div>
                                                        );
                                                    })}
                                            </div>
                                        </TableCell>
                                        <TableCell component="td" align="center" className={'startDate'}>
                                            <span title={alert.startDateLong}>{alert.startDateDisplay}</span>
                                        </TableCell>
                                        <TableCell component="td" align="center" className={'endDate'}>
                                            <span title={alert.endDateLong}>{alert.endDateDisplay}</span>
                                        </TableCell>
                                        <TableCell
                                            component="td"
                                            id={`alert-list-action-block-${alert.id}`}
                                            data-testid={`alert-list-action-block-${alert.id}`}
                                        >
                                            <AlertSplitButton
                                                alertId={alert.id}
                                                deleteAlertById={deleteAlertById}
                                                mainButtonLabel={tableType === 'past' ? 'View' : 'Edit'}
                                                navigateToCloneForm={navigateToCloneForm}
                                                navigateToEditForm={navigateToEditForm}
                                                navigateToView={navigateToView}
                                                confirmDeleteLocale={confirmDeleteLocale}
                                            />
                                        </TableCell>
                                    </TableRow>
                                );
                            })}
                        <TableRow
                            id={`alert-list-no-alerts-${tableType}`}
                            data-testid={`alert-list-no-alerts-${tableType}`}
                            style={
                                rowsPerPage > 0 && userows.length > 0 ? { display: 'none' } : { display: 'table-row' }
                            }
                        >
                            <TableCell component="td">
                                <p>No alerts</p>
                            </TableCell>
                        </TableRow>
                    </TableBody>
                    {!!needsPaginator && (
                        <TableFooter>
                            <TableRow>
                                <TablePagination
                                    rowsPerPageOptions={[5, 10, 25, { label: 'All', value: rows.length }]}
                                    colSpan={3}
                                    count={userows.length}
                                    rowsPerPage={rowsPerPage}
                                    page={page}
                                    SelectProps={{
                                        inputProps: {
                                            'aria-label': 'rows per page',
                                            'data-testid': 'admin-alerts-list-paginator-select',
                                        },
                                        native: true,
                                    }}
                                    onPageChange={handleChangePage}
                                    onRowsPerPageChange={handleChangeRowsPerPage}
                                    ActionsComponent={TablePaginationActions}
                                />
                            </TableRow>
                        </TableFooter>
                    )}
                </StyledTable>
            </TableContainer>
        </React.Fragment>
    );
};

AlertsListAsTable.propTypes = {
    rows: PropTypes.array,
    headertag: PropTypes.string,
    alertsLoading: PropTypes.any,
    actions: PropTypes.any,
    deleteAlert: PropTypes.any,
    footerDisplayMinLength: PropTypes.number,
    alertOrder: PropTypes.any,
};

AlertsListAsTable.defaultProps = {
    footerDisplayMinLength: 5, // the number of records required in the alert list before we display the paginator
    alertOrder: false, // what order should we sort the alerts in? false means unspecified
};

export default AlertsListAsTable;
