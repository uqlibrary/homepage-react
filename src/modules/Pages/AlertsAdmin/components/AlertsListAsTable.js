import React from 'react';
import PropTypes from 'prop-types';

import moment from 'moment';

import { makeStyles, useTheme } from '@material-ui/core/styles';
import Checkbox from '@material-ui/core/Checkbox';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableFooter from '@material-ui/core/TableFooter';
import TableHead from '@material-ui/core/TableHead';
import TablePagination from '@material-ui/core/TablePagination';
import TableRow from '@material-ui/core/TableRow';
import Paper from '@material-ui/core/Paper';
import IconButton from '@material-ui/core/IconButton';
import FirstPageIcon from '@material-ui/icons/FirstPage';
import KeyboardArrowLeft from '@material-ui/icons/KeyboardArrowLeft';
import KeyboardArrowRight from '@material-ui/icons/KeyboardArrowRight';
import LastPageIcon from '@material-ui/icons/LastPage';

// from https://codesandbox.io/s/hier2
// per https://material-ui.com/components/tables/#custom-pagination-actions

const useStyles1 = makeStyles(theme => ({
    root: {
        flexShrink: 0,
        marginLeft: theme.spacing(2.5),
    },
}));

function TablePaginationActions(props) {
    const classes = useStyles1();
    const theme = useTheme();
    const { count, page, rowsPerPage, onChangePage } = props;

    const handleFirstPageButtonClick = event => {
        onChangePage(event, 0);
    };

    const handleBackButtonClick = event => {
        onChangePage(event, page - 1);
    };

    const handleNextButtonClick = event => {
        onChangePage(event, page + 1);
    };

    const handleLastPageButtonClick = event => {
        onChangePage(event, Math.max(0, Math.ceil(count / rowsPerPage) - 1));
    };

    return (
        <div className={classes.root}>
            <IconButton onClick={handleFirstPageButtonClick} disabled={page === 0} aria-label="first page">
                {theme.direction === 'rtl' ? <LastPageIcon /> : <FirstPageIcon />}
            </IconButton>
            <IconButton onClick={handleBackButtonClick} disabled={page === 0} aria-label="previous page">
                {theme.direction === 'rtl' ? <KeyboardArrowRight /> : <KeyboardArrowLeft />}
            </IconButton>
            <IconButton
                onClick={handleNextButtonClick}
                disabled={page >= Math.ceil(count / rowsPerPage) - 1}
                aria-label="next page"
            >
                {theme.direction === 'rtl' ? <KeyboardArrowLeft /> : <KeyboardArrowRight />}
            </IconButton>
            <IconButton
                onClick={handleLastPageButtonClick}
                disabled={page >= Math.ceil(count / rowsPerPage) - 1}
                aria-label="last page"
            >
                {theme.direction === 'rtl' ? <FirstPageIcon /> : <LastPageIcon />}
            </IconButton>
        </div>
    );
}

TablePaginationActions.propTypes = {
    count: PropTypes.number.isRequired,
    onChangePage: PropTypes.func.isRequired,
    page: PropTypes.number.isRequired,
    rowsPerPage: PropTypes.number.isRequired,
};

const useStyles2 = makeStyles({
    table: {
        minWidth: 500,
    },
    editButton: {
        backgroundColor: '#0e62eb',
        color: '#fff',
        padding: '1em',
    },
});

export default function AlertsListAsTable(rows, dataLoading, hasFooter = false) {
    const classes = useStyles2();
    const [page, setPage] = React.useState(0);
    const [rowsPerPage, setRowsPerPage] = React.useState(5);

    const handleChangePage = (event, newPage) => {
        setPage(newPage);
    };

    const handleChangeRowsPerPage = event => {
        setRowsPerPage(parseInt(event.target.value, 10));
        setPage(0);
    };

    return (
        <TableContainer component={Paper}>
            <Table className={classes.table} aria-label="custom pagination table">
                <TableHead>
                    <TableRow md-row="" className="md-row">
                        <TableCell component="th" scope="row" />
                        <TableCell component="th" scope="row">
                            Alert
                        </TableCell>
                        <TableCell component="th" scope="row">
                            Publish Date
                        </TableCell>
                        <TableCell component="th" scope="row">
                            Unpublish Date
                        </TableCell>
                        <TableCell component="th" scope="row" />
                    </TableRow>
                </TableHead>
                <TableBody>
                    {(rowsPerPage > 0 ? rows.slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage) : rows).map(
                        alert => {
                            const startDate =
                                moment(alert.start).format('m') === '0'
                                    ? moment(alert.start).format('dddd D/MMM/YYYY ha')
                                    : moment(alert.start).format('dddd D/MMM/YYYY h.mma');
                            const endDate =
                                moment(alert.end).format('m') === '0'
                                    ? moment(alert.end).format('dddd D/MMM/YYYY ha')
                                    : moment(alert.end).format('dddd D/MMM/YYYY h.mma');
                            // Strip markdown from the body
                            const linkRegex = alert.body.match(/\[([^\]]+)\]\(([^)]+)\)/);
                            let message = alert.body;
                            if (!!linkRegex && linkRegex.length === 3) {
                                message = message.replace(linkRegex[0], '').replace('  ', ' ');
                            }
                            return (
                                <TableRow key={alert.id} id={alert.id}>
                                    <TableCell component="td">
                                        <Checkbox />
                                    </TableCell>
                                    <TableCell component="td">
                                        <b>{`${alert.title}`}</b> {`${message}`}
                                    </TableCell>
                                    <TableCell component="td" align="center">
                                        {startDate}
                                    </TableCell>
                                    <TableCell component="td" align="center">
                                        {endDate}
                                    </TableCell>
                                    <TableCell component="td">
                                        <a
                                            className={classes.editButton}
                                            data-testid="alert-list-edit"
                                            disabled={dataLoading}
                                            id="alert-list-edit-button"
                                            href={`/admin/alerts/edit/${alert.id}`}
                                        >
                                            Edit
                                        </a>
                                    </TableCell>
                                </TableRow>
                            );
                        },
                    )}
                </TableBody>
                {!!hasFooter && (
                    <TableFooter>
                        <TableRow>
                            <TablePagination
                                rowsPerPageOptions={[5, 10, 25, { label: 'All', value: -1 }]}
                                colSpan={3}
                                count={rows.length}
                                rowsPerPage={rowsPerPage}
                                page={page}
                                SelectProps={{
                                    inputProps: { 'aria-label': 'rows per page' },
                                    native: true,
                                }}
                                onChangePage={handleChangePage}
                                onChangeRowsPerPage={handleChangeRowsPerPage}
                                ActionsComponent={TablePaginationActions}
                            />
                        </TableRow>
                    </TableFooter>
                )}
            </Table>
        </TableContainer>
    );
}
