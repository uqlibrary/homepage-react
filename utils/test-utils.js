/* eslint-disable react/prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import { render } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { Route } from 'react-router';
import { mui1theme } from 'config/theme';
import { Provider } from 'react-redux';
import { ThemeProvider as MuiThemeProvider } from '@material-ui/core/styles';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';

import preview from 'jest-preview';

import { getStore } from '../src/config/store';
import Immutable from 'immutable';
import { createMemoryHistory } from 'history';

const domTestingLib = require('@testing-library/dom');
const reactTestingLib = require('@testing-library/react');

const { configure } = domTestingLib;

configure(config => ({
    ...config,
    testIdAttribute: 'id',
}));

const AllTheProviders = props => {
    return (
        <MuiThemeProvider theme={mui1theme}>
            <MuiPickersUtilsProvider utils={MomentUtils}>{props.children}</MuiPickersUtilsProvider>
        </MuiThemeProvider>
    );
};

AllTheProviders.propTypes = {
    children: PropTypes.node,
};

export const rtlRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

export const renderWithRouter = (
    ui,
    { route = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {},
) => {
    return render(
        <AllTheProviders>
            <Router history={history}>{ui}</Router>
        </AllTheProviders>,
    );
};

export const RenderWithRouter = ({
    children,
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
}) => (
    <AllTheProviders>
        <Router history={history}>{children}</Router>
    </AllTheProviders>
);
export const renderWithRedux = ({ initialState }) => render => {
    return {
        ...render,
        store: getStore({ initialState, history: render.history }),
    };
};

export const withRouter = ({
    route = '/',
    path = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
} = {}) => WrappedComponent => {
    return (
        <Router history={history}>
            <Route path={path} children={WrappedComponent} />
        </Router>
    );
};

export const withRedux = (initialState = Immutable.Map()) => WrappedComponent => {
    return <Provider store={getStore(initialState)}>{WrappedComponent}</Provider>;
};

// eslint-disable-next-line react/prop-types
export const WithRedux = ({ initialState = Immutable.Map(), children }) => (
    <Provider store={getStore(initialState)}>{children}</Provider>
);

export const WithReduxStore = ({ initialState = Immutable.Map(), children }) => (
    <Provider store={getStore(initialState)}>
        <AllTheProviders>{children}</AllTheProviders>
    </Provider>
);

module.exports = {
    ...domTestingLib,
    ...reactTestingLib,
    rtlRender,
    renderWithRouter,
    renderWithRedux,
    withRedux,
    withRouter,
    WithRedux,
    WithReduxStore,
    RenderWithRouter,
    preview,
};
