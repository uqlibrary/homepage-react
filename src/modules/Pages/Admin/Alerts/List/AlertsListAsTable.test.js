import React from 'react';
import { rtlRender } from 'test-utils';
import AlertsListAsTable from './AlertsListAsTable';
function setup(testProps = {}) {
    const props = {
        rows: [],
        headertag: 'Current alerts',
        alertsLoading: false,
        history: {},
        actions: {},
        deleteAlert: jest.fn(),
    };
    return rtlRender(<AlertsListAsTable {...props} {...testProps} />);
}

describe('defaults', () => {
    it('default values render correctly', () => {
        const { container } = setup();
        expect(container).toMatchSnapshot();
    });
});
