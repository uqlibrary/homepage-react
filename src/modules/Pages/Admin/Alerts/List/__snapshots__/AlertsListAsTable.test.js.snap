// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonalisedPanel default values render correctly 1`] = `
<div>
  <div
    class=" css-jfcehw"
    data-testid="headerRow-current"
  >
    <div>
      <h3
        style="margin-bottom: 6px;"
      >
        Current alerts
        <span
          data-testid="headerRow-count-current"
          style="font-size: 0.9em; font-weight: 300;"
        >
           
          - 
          0 alert
        </span>
      </h3>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
    data-testid="alert-list-current"
    id="alert-list-current"
  >
    <table
      aria-label="custom pagination table"
      class="MuiTable-root table css-i991eq-MuiTable-root"
      style="min-height: 200px;"
    >
      <thead
        class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head md-row css-1h2y1iv-MuiTableRow-root"
          md-row=""
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1c9afxo-MuiTableCell-root"
            scope="row"
          >
            <span
              class="screenreader"
            >
              Select
            </span>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1qewcd6-MuiTableCell-root"
            scope="row"
          >
            Alert
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1c9afxo-MuiTableCell-root"
            scope="row"
          >
            Publish date
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1c9afxo-MuiTableCell-root"
            scope="row"
          >
            Unpublish date
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1c9afxo-MuiTableCell-root"
            scope="row"
          >
            <span
              class="screenreader"
            >
              Actions
            </span>
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-1h2y1iv-MuiTableRow-root"
          data-testid="alert-list-no-alerts-current"
          id="alert-list-no-alerts-current"
          style="display: table-row;"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1xwnbnq-MuiTableCell-root"
          >
            <p>
              No alerts
            </p>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
